draft-ietf-tls-tls13.md

```diff
@@ -2089,28 +2089,42 @@ selected_group
 : The group which the client MUST use for its new ClientHello.
 {:br }
 
-The "server_version", "cipher_suite" and "extensions" fields have the
+The server_version, cipher_suite, and extensions fields have the
 same meanings as their corresponding values in the ServerHello. The
 server SHOULD send only the extensions necessary for the client to
-generate a correct ClientHello pair.
+generate a correct ClientHello pair. As with a ServerHello, only
+extensions first offered by the client can appear in a HelloRetryRequest.
 
 Upon receipt of a HelloRetryRequest, the client MUST first verify
-that the "selected_group" field does not identify a group which
-was not in the original ClientHello. If it was present, then
-the client MUST abort the handshake with a fatal "handshake_failure"
-alert. Clients SHOULD also abort with "handshake_failure" in response to any second
-HelloRetryRequest which was sent in the same connection (i.e.,
+that the selected_group was offered in its "supported_groups" extension and
+that no ClientKeyShareOffer for the selected_group was offered in its "client_key_share"
+extension. If either of these conditions are detected, then the client
+MUST abort the handshake with a fatal "handshake_failure"
+alert. Clients SHOULD also abort with an "unexpected_message" alert in response to any second
+HelloRetryRequest which is sent on the same connection (i.e.,
 where the ClientHello was itself in response to a HelloRetryRequest).
 
-Otherwise, the client MUST send a ClientHello with a new
-ClientKeyShare extension to the server. The ClientKeyShare MUST append
-a new ClientKeyShareOffer which is consistent with the
-"selected_group" field to the groups in the original ClientKeyShare.
+If the HelloRetryRequest is verified as acceptable, the client MUST send a new
+ClientHello with a new ClientKeyShare extension to the server. This ClientKeyShare MUST contain
+exactly one ClientKeyShareOffer value which is consistent with the
+selected_group field provided by the server in its HelloRetryRequest.
+All other fields in this new ClientHello MUST match the first attempt.
 
-Upon re-sending the ClientHello and receiving the
-server's ServerHello/ServerKeyShare, the client MUST verify that
-the selected CipherSuite and NamedGroup match that supplied in
-the HelloRetryRequest.
+If the new ClientHello is acceptable, the server will continue the
+handshake by replying with a new ServerHello, after which point
+the handshake follows normal procedure. (see {{incorrect-dhe-share}})
+
```

Comment:
On Saturday, October 03, 2015 10:52:46 pm ekr wrote:
> I think we crossed wires here with your note above. However, if you think of any attacks on the non-hash-continuation design, that would be very helpful. We're currently trying to get a tight analysis of that, so any thinking you have here would be most welcome.

Essentially, I was writing from a logical flow perspective without thinking about how to implement it statelessly. I missed that the keys resend was intentional to avoid the simple attack in the stateless case (issue #73, opened & closed a year ago) and I attributed #104 to covering the whole issue. Obvious in hindsight. I just thought more was incomplete here than was actually the case.

I'll hand-wave my mistake into saying that I think writing out things explicitly here will help avoid confusions and mistakes with regard to reading & implementing this. ;)

In this case, I'd actually like to just write that out explicitly here to make sure it's clear and clients don't skimp here to save bandwidth.

Will push changes shortly...